### PR TITLE
feat: Add Admin Notice for WPGraphQL for ACF

### PR DESCRIPTION
**Pull request by josephfusco**
What does this implement/fix? Explain your changes.
---------------------------------------------------

This introduces a new `register_graphql_admin_notice()` API that can be used to display an admin notice on the Plugins page, or on the WPGraphQL Settings or GraphiQL IDE page. 

This can be used by core WPGraphQL or for extensions to provide notices scoped to WPGraphQL pages.

It allows for notices to be permanent or dismissable and handles nonces for dismissing, storing a flag of dismissed notices in user_meta so that users won't be bugged again by the same notice after they've dismissed it.

Does this close any currently open issues?
------------------------------------------
no

Any other comments?
-------------------

Documentation: https://www.wpgraphql.com/functions/register_graphql_admin_notice

example usage:

```php
add_action( 'plugins_loaded', function() {

  if ( function_exists( 'register_graphql_admin_notice' ) ) {
    $config = [
      
      // required. string. The message to display in the notice. If no message, the notice won't display.
      'message' => __( 'Your notice message', 'your-textdomain' ), 
      
      // optional. boolean. whether the notice should be dismissable. False to be non-dismissable.
      'is_dismissable' => true, 
     
      // optional. string. The type of message. warning, success, error, info
      'type' => 'info', 
      
      // optional. callback. return true to display the notice (i.e. only display if another plugin is active)
      'conditions' => function() { 
         // if xx == true, return true
         // if xx == false, return false  
       }
    ];
    
    register_graphql_admin_notice( 'my-notice-slug', $config );
  }

} );
```
CleanShot 2024-02-02 at 16 58 50

### CODE REVIEW

**Comment by justlevine** [comment]
> I tested this by replacing the Admin.php file in our live on our system yesterday, and the deprecation warnings still persist today.

1. This is a _draft_ PR, so regardless it shouldn't be expected to work in its current state.

3. The purpose this PR is to display an _admin notice_ on the WordPress dashboard. It has nothing to do with #2981 or any other PHP 8.3 compatibility work.

**Comment by xbd63** [comment]
Ok thanks for explaining that.

On the WP GraphQL I do see a notice "Check out the new WPGraphQL for ACF".

Please let me know if I need to do anything else.

File: access-functions.php [hunk 1]
```diff
@@ -953,3 +953,16 @@ function str_ends_with( string $haystack, string $needle ): bool {
 		return $needle_length <= strlen( $haystack ) && 0 === substr_compare( $haystack, $needle, -$needle_length );
 	}
 }
+
+/**
+ * @param string       $slug A unique slug to identify the admin notice by
```
Comment (justlevine) [inline access-functions.php:958]:
```suggestion
 * Registers a configurable notice on the WordPress Dashboard for WPGraphQL-related notices.
 *
 * @param string       $slug A unique slug to identify the admin notice by.
```

File: access-functions.php [hunk 2]
```diff
@@ -953,3 +953,16 @@ function str_ends_with( string $haystack, string $needle ): bool {
...
+ * @param array<mixed> $config The config for the admin notice. Determines visibility, context, etc.
```
Comment (justlevine) [inline access-functions.php:959]:
```suggestion
 * @param array<string,mixed> $config The config for the admin notice. Determines visibility, context, etc.
```

File: src/Admin/Admin.php [hunk 3]
```diff
@@ -43,6 +43,9 @@ public function init() {
 		$this->admin_enabled    = apply_filters( 'graphql_show_admin', true );
 		$this->graphiql_enabled = apply_filters( 'graphql_enable_graphiql', get_graphql_setting( 'graphiql_enabled', true ) );
 
+		$admin_notices = new AdminNotices();
```
Comment (justlevine) [inline src/Admin/Admin.php:46]: This should probably be wrapped in an `is_admin()` check. I don't think we're doing that earlier in the lifecycle but I might be wrong (reviewing via 📱)

File: src/Admin/AdminNotices.php [hunk 4]
```diff
@@ -0,0 +1,299 @@
+<?php
+
+namespace WPGraphQL\Admin;
+
+/**
+ * This class isn't intended for direct extending or customizing.
```
Comment (justlevine) [inline src/Admin/AdminNotices.php:6]: We should mark the class `@internal` and not just leave it in text.

File: src/Admin/AdminNotices.php
Comment (justlevine) [inline src/Admin/AdminNotices.php:1]: Can we add a method to this for validation the config?

**Review by justlevine — COMMENTED** [review]
Did a quick pass, will review more in depth tomorrow and also probably make those array types more specific

File: composer.json [hunk 5]
```diff
@@ -47,7 +47,8 @@
 		"slevomat/coding-standard": "^8.9",
 		"szepeviktor/phpstan-wordpress": "~1.3.0",
 		"wp-graphql/wp-graphql-testcase": "~2.3",
-		"wp-cli/wp-cli-bundle": "^2.8"
+		"wp-cli/wp-cli-bundle": "^2.8",
+		"10up/wp_mock": "^0.5.0"
```
Comment (justlevine) [inline composer.json:51]: Is this PR using this library somewhere?

Comment (jasonbahl) [reply composer.json:51]: Ah! I was using it when drafting the tests but then decided not to.

File: src/Admin/AdminNotices.php [hunk 6]
```diff
@@ -0,0 +1,337 @@
...
+	}
+
+	/**
+	 * Throw an error if the config is not valid.
+	 *
+	 * @since @TODO
+	 *
+	 * @param array<string,mixed> $config The config of the admin notice
+	 *
+	 * @return bool
```
Comment (jasonbahl) [inline src/Admin/AdminNotices.php:137]:
```suggestion
	 * @param array<string,mixed> $config The config of the admin notice
```

**Pull request merged**